File: Fugazi/src/org/fugazi/ast/ASTBuilder.java
package org.fugazi.ast;

import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.fugazi.ast.form.Form;
import org.fugazi.parser.QLLexer;
import org.fugazi.parser.QLParser;

import java.io.IOException;
import java.io.InputStream;

public class ASTBuilder {

    private final QLParser parser;

    public ASTBuilder(InputStream inputStream) throws IOException {

        ANTLRInputStream input = new ANTLRInputStream(inputStream);
        QLLexer lexer = new QLLexer(input);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        parser = new QLParser(tokens);
    }

    public Form buildForm() {
        return (Form) buildFromTree(parser.form());
    }

    private AbstractASTNode buildFromTree(ParseTree tree) {
        return tree.accept(new FugaziQLVisitor());
    }
}

File: BNP/QL/QL/AST/ASTCreation/ASTBuilder.cs
using System.Linq;
using Antlr4.Runtime;
using QL.DataHandlers;
using QL.Grammar;

namespace QL.AST.ASTCreation
{
    public class ASTBuilder : IExecutableHandler
    {
        public bool Execute(DataContext context)
        {
            QLLexer lexer = new QLLexer(context.AntlrInput);
            lexer.AddErrorListener(new LexerErrorHandler(context.ASTHandlerExceptions));

            CommonTokenStream tokens = new CommonTokenStream(lexer);
            QLParser parser = new QLParser(tokens);
            parser.AddErrorListener(new ParserErrorHandler(context.ASTHandlerExceptions));

            QLListener listener = new QLListener(context.ASTHandlerExceptions);
            parser.AddParseListener(listener);

            // commence parsing the input as a formBlock since it's supposed to be the entry point of the input file
            parser.formBlock();
            context.RootNode = listener.GetAstRootNode();
            return !context.ASTHandlerExceptions.Any();
        }
    }
}
